File: app/schemas/cms.py
"""CMS admin request/response schemas."""

from datetime import datetime
from typing import Any, List, Optional

from pydantic import BaseModel, Field

from app.schemas.common import PaginatedResponse
from app.schemas.storefront import ArticleTeaser, HeroSlideResponse, ProductCommentResponse


class ArticleAdminResponse(ArticleTeaser):
    author: str
    tags: List[str] = Field(default_factory=list)
    related_product_ids: List[int] = Field(default_factory=list)
    blocks: List[dict[str, Any]] = Field(default_factory=list)
    is_published: bool = True


class ArticleCreateRequest(BaseModel):
    slug: str = Field(..., min_length=1, max_length=200)
    title: str = Field(..., min_length=1, max_length=255)
    excerpt: str = Field(..., min_length=1)
    cover_image: Optional[str] = Field(None, max_length=500)
    published_at: datetime
    reading_minutes: int = Field(default=5, ge=1)
    author: str = Field(default="تیم فنی کارزار", max_length=100)
    tags: List[str] = Field(default_factory=list)
    related_product_ids: List[int] = Field(default_factory=list)
    blocks: List[dict[str, Any]] = Field(default_factory=list)
    is_published: bool = True


class ArticleUpdateRequest(BaseModel):
    slug: Optional[str] = Field(None, min_length=1, max_length=200)
    title: Optional[str] = Field(None, min_length=1, max_length=255)
    excerpt: Optional[str] = Field(None, min_length=1)
    cover_image: Optional[str] = Field(None, max_length=500)
    published_at: Optional[datetime] = None
    reading_minutes: Optional[int] = Field(None, ge=1)
    author: Optional[str] = Field(None, max_length=100)
    tags: Optional[List[str]] = None
    related_product_ids: Optional[List[int]] = None
    blocks: Optional[List[dict[str, Any]]] = None
    is_published: Optional[bool] = None


class ArticleAdminListResponse(PaginatedResponse[ArticleAdminResponse]):
    pass


class HeroSlideCreateRequest(BaseModel):
    title: str = Field(..., min_length=1, max_length=255)
    subtitle: Optional[str] = Field(None, max_length=500)
    cta_label: Optional[str] = Field(None, max_length=100)
    cta_href: Optional[str] = Field(None, max_length=500)
    image: str = Field(..., min_length=1, max_length=500)
    accent: str = Field(default="#C22026", max_length=20)
    sort_order: int = 0
    is_active: bool = True


class HeroSlideUpdateRequest(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=255)
    subtitle: Optional[str] = Field(None, max_length=500)
    cta_label: Optional[str] = Field(None, max_length=100)
    cta_href: Optional[str] = Field(None, max_length=500)
    image: Optional[str] = Field(None, min_length=1, max_length=500)
    accent: Optional[str] = Field(None, max_length=20)
    sort_order: Optional[int] = None
    is_active: Optional[bool] = None


class HeroSlideAdminListResponse(BaseModel):
    data: List[HeroSlideResponse]


class ProductCommentCreateRequest(BaseModel):
    author_name: str = Field(..., min_length=2, max_length=100)
    rating: int = Field(..., ge=1, le=5)
    body: str = Field(..., min_length=3)
    is_verified_buyer: bool = False


class ProductCommentAdminListResponse(PaginatedResponse[ProductCommentResponse]):
    pass


class ContactSubmissionResponse(BaseModel):
    id: int
    ticket_code: str
    full_name: str
    phone: str
    subject: str
    message: str
    created_at: datetime


class ContactSubmissionListResponse(PaginatedResponse[ContactSubmissionResponse]):
    pass

File: frontend/admin-panel/src/types/cms.ts
/** CMS admin types — aligned with app/schemas/cms.py + mock-api. */

import type { PaginatedResponse } from "./common";

export const ARTICLE_BLOCK_TYPES = [
  "paragraph",
  "heading",
  "subheading",
  "list",
  "quote",
  "image",
  "table",
  "callout",
  "faq",
  "meta",
] as const;
export type ArticleBlockType = (typeof ARTICLE_BLOCK_TYPES)[number];

export const ARTICLE_BLOCK_TYPE_LABELS: Record<ArticleBlockType, string> = {
  paragraph: "پاراگراف",
  heading: "عنوان",
  subheading: "زیرعنوان",
  list: "فهرست",
  quote: "نقل‌قول",
  image: "تصویر",
  table: "جدول",
  callout: "نکته",
  faq: "سوالات متداول",
  meta: "سئو",
};

export interface ArticleBlock {
  type: ArticleBlockType;
  text?: string;
  items?: string[];
}

export interface Article {
  id: number;
  slug: string;
  title: string;
  excerpt: string;
  cover_image: string | null;
  published_at: string;
  reading_minutes: number;
  author: string;
  tags: string[];
  related_product_ids: number[];
  blocks: ArticleBlock[];
  is_published: boolean;
  created_at?: string;
  updated_at?: string;
}

export interface ArticleCreatePayload {
  slug: string;
  title: string;
  excerpt: string;
  cover_image?: string | null;
  published_at: string;
  reading_minutes: number;
  author: string;
  tags?: string[];
  related_product_ids?: number[];
  blocks?: ArticleBlock[];
  is_published: boolean;
}

export type ArticleUpdatePayload = Partial<ArticleCreatePayload>;

export interface ArticleListParams {
  skip?: number;
  limit?: number;
  search?: string;
  is_published?: boolean;
}

export type ArticleListResponse = PaginatedResponse<Article>;

export interface HeroSlide {
  id: number;
  title: string;
  subtitle: string | null;
  cta_label: string | null;
  cta_href: string | null;
  image: string;
  accent: string;
  sort_order: number;
  is_active: boolean;
}

export interface HeroSlideCreatePayload {
  title: string;
  subtitle?: string | null;
  cta_label?: string | null;
  cta_href?: string | null;
  image: string;
  accent: string;
  sort_order: number;
  is_active: boolean;
}

export type HeroSlideUpdatePayload = Partial<HeroSlideCreatePayload>;

export interface HeroSlideListParams {
  skip?: number;
  limit?: number;
  is_active?: boolean;
}

export type HeroSlideListResponse = PaginatedResponse<HeroSlide>;

export interface ProductComment {
  id: number;
  product_id: number;
  author_name: string;
  rating: number;
  body: string;
  created_at: string;
  is_verified_buyer: boolean;
}

export interface ProductCommentListParams {
  skip?: number;
  limit?: number;
  product_id?: number;
}

export type ProductCommentListResponse = PaginatedResponse<ProductComment>;

export interface ContactSubmission {
  id: number;
  ticket_code: string;
  full_name: string;
  phone: string;
  subject: string;
  message: string;
  created_at: string;
}

export interface ContactSubmissionListParams {
  skip?: number;
  limit?: number;
  search?: string;
  phone?: string;
}

export type ContactSubmissionListResponse = PaginatedResponse<ContactSubmission>;

/** Aliases used by earlier integration pages */
export type CmsArticle = Article;
export type CmsArticleCreatePayload = ArticleCreatePayload;
export type CmsArticleUpdatePayload = ArticleUpdatePayload;
export type CmsArticleListResponse = ArticleListResponse;
export type CmsHeroSlide = HeroSlide;
export type CmsHeroCreatePayload = HeroSlideCreatePayload;
export type CmsHeroUpdatePayload = HeroSlideUpdatePayload;
export type CmsProductComment = ProductComment;
export type CmsCommentListResponse = ProductCommentListResponse;
export type CmsContactSubmission = ContactSubmission;
export type CmsContactListResponse = ContactSubmissionListResponse;
